apmproxy/apmserver.go

```diff
@@ -60,11 +62,17 @@ func (c *Client) ForwardApmData(ctx context.Context) error {
 			return nil
 		case data := <-c.AgentDataChannel:
 			if err := c.forwardAgentData(ctx, data); err != nil {
+				if errors.Is(err, accumulator.ErrNoData) {
+					c.logger.Debugf("received something from '%s' without APMData", data.AgentInfo)
+					continue
+				}
 				return err
 			}
-			// Wait for metadata to be available, metadata will be available as soon as
-			// the first agent data is processed.
-			lambdaDataChan = c.LambdaDataChannel
+			once.Do(func() {
```

Comment:
Not sure what's happening here, why don't we use `c.LambdaDataChannel` directly and have to first assign it to this local variable? Isn't this a NOP ?

Reply:
@kruskall or @vishal-jogi maybe can provide more insights on this design decission in the first place.
With `sync.once` the local variable should be initialized exactly once instead of every time with some received from `c.AgentDataChannel`. I just have learned, that `c.LambdaDataChannel` should not change over time.

Reply:
I don't see why we require a sync primitive here. There is no race: `lambdaDataChan` is a local variable and `c.AgentDataChannel` is created only once. If we really want to do the initialization of the local variable only once then we can just do a nil check or something.

> why don't we use c.LambdaDataChannel directly and have to first assign it to this local variable? Isn't this a NOP ?

We don't use the `c.LambdaDataChannel` because we don't want to consume from the lambda API until we are sure we can process the data and send it to APM server -- this means that we have the metadata available from the APM agents. This allows us to pushback to the lambda API rather than consuming data without metadata being available and losing the data.

Reply:
Ok makes sense.

Reply:
Agreed, I would recommend `nil` check because there is no need for a `sync` here

Reply:
I have updated the comment. Hope it clarifies the assumptions and expectations at this point.

Reply:
Hm, I still don't understand why a `nil` check is not sufficient here.

Reply:
Updated the PR to use a `nil` check instead of `sync.Once()` with https://github.com/elastic/apm-aws-lambda/pull/570/commits/1763b69bc212ca52f5c8942eb2359b2e15cbbfec.